src/level_four/tap_cards/controller/pair_card_controller.c

```diff
@@ -140,48 +152,82 @@ static void _tap_card_backend(uint8_t card_number) {
              invalid_self_keypath,
              sizeof(invalid_self_keypath)) == 0) {
     /* Device is not provisioned */
-    reset_flow_level();
-    instruction_scr_destructor();
-    mark_error_screen(ui_text_device_compromised);
-    return;
+
+    display_error_message(ui_text_device_compromised);
+    return false;
   }
 
+  /// Sign pairing data and append signature
   sha256_Raw(card_pairing_data, pairing_data_len, digest);
   uint8_t status = atecc_nfc_sign_hash(digest, sig);
   if (status)
     LOG_CRITICAL("xxec %d:%d", status, __LINE__);
   pairing_data_len +=
       ecdsa_sig_to_der(sig, card_pairing_data + pairing_data_len);
-  tap_card_data.retries = 5;
+  card_data.nfc_data.retries = 5;
 
   while (1) {
-    tap_card_data.acceptable_cards = (1 << (card_number - 1));
-    tap_card_data.lvl3_retry_point = flow_level.level_three - 1;
-    memcpy(tap_card_data.family_id, get_family_id(), FAMILY_ID_SIZE);
-    if (nfc_wait_for_card(DEFAULT_NFC_TG_INIT_TIME) != STM_SUCCESS)
-      instruction_scr_change_text(ui_text_card_removed_fast, true);
-    if (!tap_card_applet_connection())
-      return;
-    if (is_paired(tap_card_data.card_key_id) > -1) {
-      instruction_scr_destructor();
-      flow_level.level_four++;
-      return;
+    // Initialize card tap config
+    card_data.nfc_data.acceptable_cards = (1 << (card_number - 1));
+    card_data.nfc_data.lvl3_retry_point = flow_level.level_three - 1;
+    memcpy(card_data.nfc_data.family_id, get_family_id(), FAMILY_ID_SIZE);
+
+    // Initialize card applet
+    card_initialize_applet(&card_data);
+
+    if (CARD_OPERATION_SUCCESS == card_data.error_type) {
+      // if card tapped is not paired, proceed with pairing
+      if (-1 == is_paired(card_data.nfc_data.card_key_id)) {
+        card_data.nfc_data.status =
+            nfc_pair(card_pairing_data, &pairing_data_len);
+
+        if (card_data.nfc_data.status == SW_NO_ERROR) {
+          if (*(uint32_t *)get_family_id() == DEFAULT_UINT32_IN_FLASH)
+            set_family_id_flash(card_data.nfc_data.family_id);
+          handle_pair_card_success(
+              card_number, session_nonce, card_pairing_data);
+        } else {
+          card_handle_errors(&card_data);
+        }
+      }
+    }
+
+    // Display error message if set during card initialization or operation
+    // error handling
+    if (NULL != card_data.error_message) {
+      buzzer_start(BUZZER_DURATION);
+      if (CARD_OPERATION_SUCCESS !=
+          display_error_message(card_data.error_message)) {
+        result = false;
+        break;
+      }
+
+      // Reset instruction screen if retap required
+      if (CARD_OPERATION_RETAP_BY_USER_REQUIRED == card_data.error_type) {
+        instruction_scr_init(message, heading);
```

Comment:
Let this be the first statement inside while(1) loop so that we don't need to worry about this.. in case any error occurs and we have to retry we will not have to worry..

Saying because we missed calling this in line 221,222,223 where retap was required

Reply:
We want to set the instruction screen on start of the flow and in while(1) when recoverable error is set.

For this requirement, we can add the instruction screen just before entry to while loop.

Reply:
updated